Automate air filtration with air quality sensors (#22)
viam-labs/viamcodelabs: site/viamguides/src/air-quality/air-quality.md

  ![box fan and air filter](assets/boxFanClips.jpg)

### Set up your smart plug
1. Plug your Kasa smart plug into a power outlet. 
1. Set up your smart plug and connect it to your local Wifi [using the Kasa mobile app](https://www.tp-link.com/us/support/faq/946/) from the App Store or Google Play. 
1. Find your smart plug's IP address on your router's admin page. From a web browser, enter your router's IP address (commonly `192.168.1.1` or `192.168.0.1`) and log in using your admin username and password. Look for a section like "Connected Devices," "Device List," or "DHCP Clients." Locate your Kasa smart plug by its name, MAC address (shown in the Kasa mobile app), or manufacturer name (usually TP-Link). Make a note of the IP address since we'll need it again soon.
    > aside negative
    > **Command line alternative** install the [`python-kasa`](https://python-kasa.readthedocs.io/en/stable/index.html) library, and enter the command `kasa discover` to locate the IP address of the connected device.
1. Plug the box fan into the Kasa smart plug. 
  ![photo of Kasa smart plug with fan](assets/smartPlug.jpg)

<!-- ------------------------ -->
## Set up your Raspberry Pi
Duration: 6

> The Raspberry Pi boots from a USB flash drive (or microSD card). You need to install Raspberry Pi OS on a USB flash drive that you will use with your Pi. For more details about alternative methods of setting up your Raspberry Pi, refer to the [Viam docs](https://docs.viam.com/installation/prepare/rpi-setup/#install-raspberry-pi-os).

Author: I agree that it helps to keep things modular. I find it helpful to go through areas tip-to-tail, and included some details that aren't in our standard docs. 

In the future, if I created a "Set up viam-server on Raspberry Pi 5" codelab or something like that, I am making a mental note to swap that out here as well.

1. Connect your Pi to Wi-Fi so that you can run `viam-server` wirelessly. Check **Configure wireless LAN** and enter your wireless network credentials. SSID (short for Service Set Identifier) is your Wi-Fi network name, and password is the network password. Change the section `Wireless LAN country` to where your router is currently being operated.
  ![raspberry pi hostname username and password](assets/generalSettings.png)
1. Select the **SERVICES** tab, check **Enable SSH**, and select **Use password authentication**.
    ![raspberry pi enable SSH](assets/sshSettings.png)
    > aside negative
    > Be sure that you remember the `hostname` and `username` you set, as you will need this when you SSH into your Pi.
1. **Save** your updates, and confirm `YES` to apply OS customization settings. Confirm `YES` to erase data on the USB flash drive. You may also be prompted by your operating system to enter an administrator password. After granting permissions to the Imager, it will begin writing and then verifying the Linux installation to the USB flash drive.
1. Remove the USB flash drive from your computer when the installation is complete.

### Connect with SSH

> 1. Place the USB flash drive into your Raspberry Pi and boot the Pi by plugging it in to an outlet. A red LED will turn on to indicate that the Pi is connected to power.

Reviewer: I like the idea of using a USB flash drive by default over an SD card for longevity and availability (most folks probably have one or two lying around). We should be sure to call out anything that could prevent booting from USB working out of the box, like using a lower amperage power supply with the Raspberry Pi 5:

> USB boot is disabled by default when connected to a 3A power supply. 
https://www.raspberrypi.com/documentation/computers/raspberry-pi.html#differences-on-raspberry-pi-5

The power supply that should be included by Raspberry Pi is 5V 5A (25W). However, folks who have owned a Pi before or got an unofficial power supply might be using 5V 3A (15W) since it was the default previously. 

I know you called out the need for 5V in a previous step, but we should probably include the 5A (amps) requirement for optimal performance too.

    ```
1. Use your keyboard to select “Interface Options”, and press return.
    ![raspi config](assets/enablePeriph.png)
1. [Enable the relevant protocols](https://docs.viam.com/installation/prepare/rpi-setup/#enable-communication-protocols) to support our hardware. Since you are using a sensor that communicates over the serial port, enable **Serial Port**.
  ![enable serial](assets/enableSerial.png)
1. Confirm the options to enable the serial login shell and serial interface. And reboot the Pi when you're finished.

<!-- ------------------------ -->
## Configure your machine and peripherals
Duration: 5

### Configure your machine

1. In [the Viam app](https://app.viam.com/robots), create a machine by typing in a name and clicking **Add machine**. 
  ![add machine](assets/addMachine.png)
1. Click **View setup instructions**.
  ![setup instructions](assets/awaitSetup.png)
1. Install `viam-server` on the Raspberry Pi device that you want to use to communicate with and control your air sensor. Select the `Linux / Aarch64` platform for the Raspberry Pi to control the air sensor, and leave your installation method as `viam-agent`.
  ![select platform](assets/selectPlatform.png)

> 1. Use the `viam-agent` to download and install `viam-server` on your Raspberry Pi. Follow the instructions to run the command provided in the setup instructions from the SSH prompt of your Raspberry Pi.

Reviewer: Should we be linking out the docs page on viam-agent? Not sure if it's confusing to say we're installing viam-agent in order to get viam-server running on the device.

Author: Added link - definitely confusing. But both are mentioned on that setup page for now.

  ![installation agent](assets/installAgent.png)
  The setup page will indicate when the machine is successfully connected. 

### Add your Raspberry Pi
1. In [the Viam app](https://app.viam.com/robots), find the **CONFIGURE** tab. It's time to configure your hardware.
1. Click the **+** icon in the left-hand menu and select **Component**. 
  ![select component](assets/selectComponent.png)

> 1. Select `board`, and find the `pi5` module. This adds the module for working with the Raspberry Pi 5. Leave the default name `board-1` for now.

Reviewer: We should be clear that this module is necessary for using the GPIO pins, not for running anything on the device:

```suggestion
1. Select `board`, and find the `pi5` module. This adds the module for working with the Raspberry Pi 5's GPIO pins. Leave the default name `board-1` for now.
```

1. Notice adding this module adds the board hardware component called `board-1`. The collapsible card on the right corresponds to the part listed in the left sidebar. 
  ![added board](assets/addedBoard.png)
1. Click **Save** in the top right to save and apply your configuration changes.
    > aside negative
    > If any problems occur, check under the **LOGS** tab to see what might be going wrong.

### Add your air sensor

1. In [the Viam app](https://app.viam.com/robots), click the **+** icon and select **Component**. Select `sensor`, find the `air:pms7003` module, and click **Add module**. This module provides the sensor model that supports the specific hardware we are using for this tutorial. Leave the default name `sensor-1` for now.
  ![select air sensor module](assets/selectAirSensor.png)
1. Notice adding this module adds the sensor hardware component called `sensor-1` as well as a module called `pms7003` which contains the logic for how our Raspberry Pi and air sensor work with each other. 
  ![view after adding sensor](assets/addedSensor.png)
1. **Save** your updates.

### Add your smart plug

> 1. In [the Viam app](https://app.viam.com/robots), let's configure the rest of our hardware so that your machine can communicate with the smart plug over wifi. Click the **+** icon in the left-hand menu and select **Component**. Select `generic`, and find the `kasa:switch` module. This is a generic module that's been configured to use the `python-kasa` library. In other words, it's a layer of abstraction that lets us use the Kasa manufacturer's APIs with Viam.

Reviewer: > In other words, it's a layer of abstraction that lets us use the Kasa manufacturer's APIs with Viam

Love this 👏

    ```json
    {
      "toggle_off": []
    }
    ```
    ![do command](assets/doCommand.png)
1. Now update `"toggle_off"` to `"toggle_on"` in the JSON object, and **Execute** once again.
1. From [the module listing in the registry](https://app.viam.com/module/joyce/kasasmartplug), you may be able to refer to additional documentation in [the module's README](https://github.com/loopDelicious/viam-kasa-switch) to learn about other functions you can use to test your hardware.

<!-- ------------------------ -->
## Program your air monitoring device
Duration: 6

At this point, you have configured and tested your machine and peripherals, but nothing is happening automatically. In the next section, program your air monitoring device to be a little smarter. Make your machine periodically get readings, turn on the air filter if the values are high, and turn off again when the values are low.

### Create an automation script

> 1. To configure the machine to automatically run a command to execute a script, use a [Viam process](https://docs.viam.com/configure/processes/). Create a new file called `process.py`.

Reviewer: We should probably be specific about where this new file should be created since the reader might still be ssh'd into their Pi:

```suggestion
1. To configure the machine to automatically run a command to execute a script, use a [Viam process](https://docs.viam.com/configure/processes/). Create a new file on your computer called `process.py`.
```

    ```bash
    $ touch process.py
    ```
1. Copy and paste [this sample code](https://github.com/loopDelicious/viam-pm25-process/blob/main/process.py) into the new file `process.py`. This code will allow your Raspberry Pi to connect to both our sensor and plug and execute our logic.
1. Now it's time to move your control code to your Raspberry Pi device. [SSH into your Raspberry Pi](https://docs.viam.com/installation/prepare/rpi-setup/#connect-with-ssh) if you're not already SSH'd.
1. From the SSH prompt on your Raspberry Pi, create a directory called `process` to hold your code. And then, change into that directory.
    ```bash
    $ mkdir process
    $ cd process
    ```
1. Install the Python package manager.
    ```bash
    $ sudo apt install python3-pip
    ```
1. Install the Viam Python SDK **into that new directory**.
    ```bash

>     $ pip3 install --target=process viam-sdk

Reviewer: If the reader changes directories to be in the `process` directory, passing the `--target` flag will create a new subdirectory called `process`, in which pip will install the `viam-sdk` with its dependencies, so we could probably shorten this up to remove the first step of making the `process` directory and rely on pip to do it:

```markdown
1. Install the Python package manager.
    ```bash
    $ sudo apt install -y python3-pip
    ```
1. Install the Viam Python SDK into a new directory called `process`.
    ```bash
    $ pip3 install --target=process viam-sdk
    ```
```

As an aside, I added the `-y` as part of the `apt install` command to automatically approve the install rather than requiring the reader to hit `y/n` as part of the interactive command.

  ![configure process](assets/configureProcess.png)
1. Still within the `process-1` card, select the advanced settings icon near the top right corner to review the configuration JSON. Create a new `env` property, and add your environment variables within the new property, formatted like the following with your own credentials. 
    ```json
      "env": {
        "SENSOR_NAME": "sensor-1",
        "PLUG_NAME": "generic-1",
        "ROBOT_API_KEY": "your-api-key",
        "ROBOT_API_KEY_ID": "your-api-key-id",
        "ROBOT_ADDRESS": "your-robot-address"
      },
    ```
    ![configure JSON](assets/configJSON.png)
    > aside negative
    > The `SENSOR_NAME` and `PLUG_NAME` are the default names for our air sensor and smart plug when added to our Viam machine. Other machine credentials can be found under the **CONNECT** tab, selecting an SDK, and toggling **Include API key** to reveal your credentials within the code sample.
        ![get credentials](assets/apiKey.png)
1. **Save** your updates.

> 1. Restart the machine.

Reviewer: Was it necessary to restart viam-server in order for the process to start? It _should_ happen as soon as the process configuration has been saved. 

Also, the "Restart" button is restarting the server not the device, which might be confusing if we say "restart the machine":

```suggestion
1. Restart viam-server.
```
